sklearn/tests/test_metaestimators_metadata_routing.py

```diff
@@ -628,14 +645,14 @@ def test_error_on_missing_requests_for_sub_estimator(metaestimator):
                 set_requests(
                     estimator,
                     method_mapping=metaestimator.get("method_mapping", {}),
-                    methods=["fit"],
+                    methods=[method_name],
                     metadata_name=key,
                     value=None,
                 )
-                try:
-                    # `fit` and `partial_fit` accept y, others don't.
+                if method_name in ["fit", "partial_fit", "score"]:
+                    # `fit`, `partial_fit`, 'score' accept y, others don't.
                     method(X, y, **method_kwargs)
-                except TypeError:
+                else:
```

Comment:
This try/except pattern makes it very hard to debug the errors and is not explicit. This pattern makes it explicit which methods use `y` and which do not.

Reply:
Are these the only three methods that can expect y across all estimators in the codebase?

Reply:
Yes at least within the metadata testing framework. One can add more as they arise?

Reply:
I think with try, except we don't really need to bother to maintain this list. However I am fine with either way. I'll let @adrinjalali finalise.

Reply:
I'm not sure why this makes things hard to debug really. The try/except is more foolproof since a library like imbalance-learn adds some methods to the whole system by patching a few things in sklearn and things would just work.